### PR TITLE
feat: speed conversion

### LINKED ISSUE #1423: [BUG] Incorrect Mach/airspeed conversion
**Mod Version**
master
built: 2020-10-17T02:56:44+00:00
ref: refs/heads/master
sha: 8675df0eedca8320e8f22ed1aa50bb4ba8db2a60
actor: Benjozork
event_name: push

**Describe the bug**
Mach and airspeed in KCAS have a defined relationship based on pressure altitude and temperature. When flying the A320, the relationship between these values appears to be incorrect. For example, at 29,000 feet pressure altitude at ISA conditions, any Mach-KCAS calculator or chart will show that M.078 corresponds to an airspeed of 302 KCAS (see first screenshot attached (145)). However, in the A320, at 302 KCAS, the speed bug on PFD appears to be placed around M0.76 or M0.77, and the speed the autopilot settled on was M0.764 (see 2nd screenshot - 130).

When the Mach/Airspeed is changed to Mach and 0.78 is selected, the airspeed is set to 311 KCAS, not 302 KCAS. See the next 2 screenshots showing the PFD speed bug at about 311 KCAS when 0.78 is in the FCU and indicated on the PFD (screenshot 131) and when the Mach/Airspeed setting is switched back to Airspeed without changing the set speed, it indeed shows 311 KCAS (screenshot 132).

The next set of screenshots show the same issue at 39,000 feet pressure altitude, ISA. At M0.78, the airspeed should be 241 KCAS (screenshot 146). At 241, knots, the Mach shown on the PFD is 0.773 (screenshot 133). Switching to Mach in the FCU without changing the speed setting, the FCU indeed reads 0.77 (screenshot 134). With 0.78 Mach set in the FCU, the airspeed bug on the PFD appears to be set at about 244 KCS instead of the correct 241 KCAS (screenshot 135). And, indeed, when switching the FCU to display airspeed without changing the speed setting, it shows 144 KCAS (screenshot 136).
Screenshot (145)
Screenshot (130)
Screenshot (131)
Screenshot (132)
Screenshot (146)
Screenshot (133)
Screenshot (134)
Screenshot (135)
Screenshot (136)

**To Reproduce**

1. Set weather setting to clear and calm, ISA conditions. 
2. At a weight for which the A320 can easily reach 39,000 feet at M0.78 (I used the default of around 134,000 lb), takeoff and climb to 29,000 feet. Verify from a calculator or chart that 302 KCAS is M0.78 at 29,000 feet, ISA conditions. 
3. Have Mach/airspeed switch set to airspeed, set an airspeed of 302 knots in the FCU, and capture that speed using the autopilot/autothrust. 
4. Note what Mach number is displayed in the PFD when stabilized. 
5. Without changing the airspeed setting, switch the Mach/airspeed display to Mach. Note Mach number displayed in the FCU. 
6. If desired, repeat the above steps after climbing to 39,000 feet (or any altitude for that matter).

**Expected behavior**
KCAS and Mach displayed on the PFD and selected in the FCU should agree with the known Mach-KCAS relationship

**Actual behavior**
For my flight described above, the KCAS and Mach did not follow the known relationship to each other.

**References**
see above

**Additional context**

Was this working before/when did the issue start occurring?No, it is a problem in the vanilla game. I have submitted a Zendesk report, but thought I would submit one here as well if you all have a means to fix it.

Is this a problem in the vanilla unmodded game?Yes

Discord username (if different from GitHub):donbikes#4084

### PR DESCRIPTION
**Pull request by MisterChocker**
Fixes #1423

## Summary of Changes

- Changed speed sources from `Simplane.getIndicatedSpeed()` to new simvar `L:A32NX_SPEEDS_KCAS`
- Added correction for athr commanded speed, fpd speed bugs and fcu speed indication

## Additional context

We've found out that the simulator IAS calculation is wrong.
Read more in [Is_IAS_indicated_wrongly_by_FS_.pdf](https://github.com/flybywiresim/a32nx/files/5744000/Is_IAS_indicated_wrongly_by_FS_.pdf)

Discord username (if different from GitHub): DerL30N#3751
[True_Airspeed_to_Calibrated_Airspeed_Conversion.pdf](https://github.com/flybywiresim/a32nx/files/5744001/True_Airspeed_to_Calibrated_Airspeed_Conversion.pdf)

## Testing instructions

Use both managed and selected speed modes and check if the athr targets and holds the target speeds.
(should be no better than before this pr)

Also please check the PFD indicating the correct speeds. [Speed conversion tool](https://aerotoolbox.com/airspeed-conversions/)

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by wpine215** [comment]
#2781 has been merged

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/A320_Neo_PFD.html [hunk 1]
```diff
@@ -102,11 +102,13 @@
     </div>
 </script>
 
+<script type="text/javascript" src="/JS/debug.js"></script>
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/A320_Neo_PFD.html:105]:
```suggestion
```

**Comment by Benjozork** [comment]
What's the status on this? @MisterChocker

**Comment by MisterChocker** [comment]
Integration of this is difficult due to AP/FCU limitations.
I've put this on hold due to the fact that asobo is looking into their speed calculation.
If needed I would like to integrate this in the custom ap (as integration with the default ap is very difficult)

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js [hunk 2]
```diff
@@ -8,20 +8,22 @@ class A32NX_Speeds {
...
-        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", 0);
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VS", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VLS", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_F", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_S", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_GD", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_LANDING_CONF3", "boolean", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VMAX", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", 0).then();
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_KCAS", "number", 0).then();
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js:19]:
```suggestion
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VS", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VLS", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_F", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_S", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_GD", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_LANDING_CONF3", "boolean", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VMAX", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", 0);
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_KCAS", "number", 0);
```

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js [hunk 3]
```diff
@@ -43,24 +45,37 @@ class A32NX_Speeds {
...
-            SimVar.SetSimVarValue("L:A32NX_SPEEDS_GD", "number", speeds.gd);
-            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VMAX", "number", speeds.vmax);
-            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", speeds.vfeN);
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VS", "number", speeds.vs).then();
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VLS", "number", speeds.vls).then();
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_F", "number", speeds.f).then();
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_S", "number", speeds.s).then();
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_GD", "number", speeds.gd).then();
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VMAX", "number", speeds.vmax).then();
+            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", speeds.vfeN).then();
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js:60]:
```suggestion
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VS", "number", speeds.vs);
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VLS", "number", speeds.vls);
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_F", "number", speeds.f);
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_S", "number", speeds.s);
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_GD", "number", speeds.gd);
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VMAX", "number", speeds.vmax);
            SimVar.SetSimVarValue("L:A32NX_SPEEDS_VFEN", "number", speeds.vfeN);
```

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js [hunk 4]
```diff
@@ -43,24 +45,37 @@ class A32NX_Speeds {
...
+        if (tas === this.tas && tat === this.tat && p === this.p) {
+            return;
+        }
+
+        this.tas = tas;
+        this.tat = tat;
+        this.p = p;
+
+        const kcas = NXSpeedsUtils.getKCasFromTas(tas, tat, p);
+        SimVar.SetSimVarValue("L:A32NX_SPEEDS_KCAS", "number", kcas).then();
```
Comment (MisterChocker) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js:78]:
```suggestion
        SimVar.SetSimVarValue("L:A32NX_SPEEDS_KCAS", "number", kcas);
```

**Comment by MisterChocker** [comment]
Updated the branch for easier cap integration.

**Comment by MisterChocker** [comment]
Closing this since it's going into CAP only.

**Pull request closed without merging**